### PR TITLE
Update instructions

### PR DESCRIPTION
**Pull request by rhawalsh**
This PR establishes the initial set of instructions for actually *using* the bunsen code.

Signed-off-by: Andrew Walsh <awalsh@redhat.com>

### CODE REVIEW

**Comment by sweettea** [comment]
Looks good to me, fwiw. Thank you for adding instructions :)

**Comment by rhawalsh** [comment]
I am also hoping to have a bare bones set of instructions to do basic development as well, this week.  Kinda hard to use the environment.  Stay tuned.

File: Getting_Started.md [hunk 1]
```diff
@@ -0,0 +1,128 @@
+# Getting Started with Bunsen
+
+This guide is intended to help you get your first Bunsen environment up and running.
+
+# Expectations
+
+This project is simply an ansible playbook that has a certain set of requirements that need to be satisfied moving in.
```
Comment (jshimkus-rh) [inline Getting_Started.md]:
```suggestion
This project is an ansible playbook with specific requirements that need to be satisfied before beginning.
```

File: Getting_Started.md [hunk 2]
```diff
@@ -0,0 +1,128 @@
+# Getting Started with Bunsen
+
+This guide is intended to help you get your first Bunsen environment up and running.
```
Comment (jshimkus-rh) [inline Getting_Started.md]:
```suggestion
This guide is intended to help get your Bunsen environment up and running.
```

File: Getting_Started.md [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+
+# Expectations
+
+This project is simply an ansible playbook that has a certain set of requirements that need to be satisfied moving in.
+
+## Initial requirements
+
+Before you can start provisioning your environment, the following items must be satisfied:
```
Comment (jshimkus-rh) [inline Getting_Started.md]:
```suggestion
Before you can start provisioning your environment, the following criteria must be satisfied:
```

File: Getting_Started.md [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+
+1. An environment must consist of at least 3 different systems (physical or virtual).  It is recommended that you have at least 5 machines, but minimally 3 machines can work.  More information on what the different systems will be doing will be provided later.
```
Comment (jshimkus-rh) [inline Getting_Started.md]:
```suggestion
1. At least 3 individual systems, physical and/or virtual.
    - It is recommended to have at least 5 systems.
    - The function of the separate systems are described below.
```

File: Getting_Started.md [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+    - Of those 3 machines, one must have at least 280GiB of additional storage in a separate block device (i.e. sdb)
```
Comment (jshimkus-rh) [inline Getting_Started.md]:
```suggestion
    - At least one of the 3 systems must have a minimum of 280GiB of additional storage in a separate block device.
      - Generally this is /dev/sdb.
      - If the system is virtualized using 'kvm' it must be /dev/vdb.
```

File: Getting_Started.md [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+                     -l resources \
+                     provisioning/playbook.yml
+
+    ansible-playbook -i inventory \
+                     -l farms \
+                     provisioning/playbook.yml
+    ```
+
+    - Note the added options `--extra-vars="install_bunsen_support=true" --ask-become-pass` on the infrastructure role.
+        - `--extra-vars="install_bunsen_support=true"` instructs the ansible playbook to install the necessary COPR repositories on the `controller` system and install the python dependencies (python3-resource-discovery, and python3-utility-mill).  It is only needed to be specified one time.
```
Comment (slegendr) [inline Getting_Started.md]: Line 93: I'd suggest replacing the last sentence with "It only needs to be specified one time".

File: Getting_Started.md [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+        - `--ask-become-pass` instructs the `ansible-playbook` command to prompr for the sudoers password on the `controller` to allow for you to install the packages.
```
Comment (slegendr) [inline Getting_Started.md]: Line 94: 'prompr' should be 'prompt'

File: Getting_Started.md [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+
+This project is simply an ansible playbook that has a certain set of requirements that need to be satisfied moving in.
+
+## Initial requirements
+
+Before you can start provisioning your environment, the following items must be satisfied:
+
+1. An environment must consist of at least 3 different systems (physical or virtual).  It is recommended that you have at least 5 machines, but minimally 3 machines can work.  More information on what the different systems will be doing will be provided later.
+    - Of those 3 machines, one must have at least 280GiB of additional storage in a separate block device (i.e. sdb)
+2. Each machine must be able to communicate with the others in the environment both by IP address and by name (i.e. ossbunsen-resource [10.2.0.5] needs to be able to ssh to ossbunsen-farm-f36-1 [10.2.0.6])
```
Comment (jshimkus-rh) [inline Getting_Started.md]:
```suggestion
2. Each system must be able to communicate with the others in the environment both by IP address and by name (e.g., ossbunsen-resource [10.2.0.5] needs to be able to ssh to ossbunsen-farm-f36-1 [10.2.0.6])
```

File: Getting_Started.md [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+        - `--ask-become-pass` instructs the `ansible-playbook` command to prompr for the sudoers password on the `controller` to allow for you to install the packages.
+        - If you would prefer NOT letting ansible provision your `controller`, please see the [Manually Installing Bunsen Support Libs](#installing-bunsen-support-libraries)
+      
+# Using the Bunsen environment
+
+While the environment is yours and you can make it work as you wish, the most effective workflow involves doing general development on the `resource` machine and then doing the compilations and testing from that machine.
+
+The home directory for the `target_account` specified above is mounted via NFS, so if you are doing work that requires multiple distributions, you don't have to unnecessarily duplicate things.
+
+To perform general development and testing for VDO, please see the VDO Getting Started document (This does not exist yet, contents TBD)
```
Comment (slegendr) [inline Getting_Started.md:108]: Line 103: Is this referring to this Getting_started.md, or something else? It's a bit confusing.

File: Getting_Started.md (hunk 1 above)
Comment (slegendr) [inline Getting_Started.md]: I'm getting stuck on 'moving in'. Maybe change to 'before proceeding'?

**Review by slegendr — CHANGES_REQUESTED** [review]
Overall, looks good. A few minor suggestions.

File: Getting_Started.md [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+    - Of those 3 machines, one must have at least 280GiB of additional storage in a separate block device (i.e. sdb)
+2. Each machine must be able to communicate with the others in the environment both by IP address and by name (i.e. ossbunsen-resource [10.2.0.5] needs to be able to ssh to ossbunsen-farm-f36-1 [10.2.0.6])
+3. The provisioning user must have an ssh-key generated and located under `~/.ssh`.
+4. A `controller` system that has ansible installed on it.
+
+You can find an example environment that utilizes vagrant and libvirt at [the VDO vagrant example repository](https://github.com/dm-vdo/vagrant)
+
+## System Functions
+
+A provisioned Bunsen Environment consists of three different types of machine.
```
Comment (jshimkus-rh) [inline Getting_Started.md]:
```suggestion
A provisioned Bunsen Environment consists of three different classes of system.
```

File: Getting_Started.md [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+
+- Infrastructure
+    - This machine will provide necessary infrastructure services for the other systems to utilize.
```
Comment (jshimkus-rh) [inline Getting_Started.md]:
```suggestion
    - This system provides necessary services for the other systems to utilize.
```

File: Getting_Started.md [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+        - Such as NFS, [RSVP](https://github.com/dm-vdo/permabit-rsvpd) (resource management and coordination), iSCSI backing storage (if applcable)
+- Resource
+    - General development and compiling will be done on this machine.
```
Comment (jshimkus-rh) [inline Getting_Started.md]:
```suggestion
    - General development and compiling will be done on these systems.
```

File: Getting_Started.md [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+    - If there are multiple incompatible targets (i.e. Fedora, RHEL), then one of each possible target should be available to ensure that the compilations are done according to their environment.
```
Comment (jshimkus-rh) [inline Getting_Started.md]:
```suggestion
    - If there are multiple incompatible targets (e.g., Fedora and RHEL), then one of each possible target should be available to ensure that the compilations are done according to their environment.
```

File: Getting_Started.md [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+- Farm
+    - At least one of each applicable targets (i.e. Fedora, RHEL).
```
Comment (jshimkus-rh) [inline Getting_Started.md]:
```suggestion
    - At least one for each applicable target (e.g., Fedora or RHEL).
```

File: Getting_Started.md [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+        - With Fedora, we typically utilize at least one of each supported version of Fedora in case there is a change applied to one and not the other.
+    - Each farm must have an additional disk (i.e. sdb) with 280 GiB or more of storage to use as scratch space for testing.
```
Comment (jshimkus-rh) [inline Getting_Started.md]:
```suggestion
    - Each farm must have an additional disk (see `Initial Requirements` for size and naming requirements) to use as scratch space for testing.
```

File: Getting_Started.md [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+- Controller
+    - This system can be the hypervisor that is running the various VMs.  Or it could be an additional VM living next to the others.
+
+# Deploying your first environment
+
+Assuming that the dependencies above have been satisfied, you can now provision a Bunsen environment.  In order to deploy a bunsen environment, you need a few things first.
```
Comment (jshimkus-rh) [inline Getting_Started.md]:
```suggestion
Assuming that the dependencies above have been satisfied, you can now provision a Bunsen environment.  In order to deploy a bunsen environment, you need the following...
```

File: Getting_Started.md [hunk 17]
```diff
@@ -0,0 +1,128 @@
...
+
+1. First, we need a copy of this repository placed on a `controller` somewhere.
```
Comment (jshimkus-rh) [inline Getting_Started.md]:
```suggestion
1. A copy of this repository local to a `controller` system
```

File: Getting_Started.md [hunk 18]
```diff
@@ -0,0 +1,128 @@
...
+    - This system is not part of the minimal environment specified above.
+2. Next, we need to generate an inventory file for ansible to utilize when provisioning the hosts.
```
Comment (jshimkus-rh) [inline Getting_Started.md]:
```suggestion
2. An inventory file for ansible to utilize when provisioning the hosts
```

File: Getting_Started.md [hunk 19]
```diff
@@ -0,0 +1,128 @@
...
+    ossbunsen-farm-f36-1
+    
+    [all:vars]
+    deploying_account = admin
+    target_account = admin
+    ```
+
+    The `admin` account should be replaced with the account that you intend to do your development with.
+
+    NOTE: In place of `ansible_ssh_pass=vagrant`, you can use `ansible_ssh_private_key_file='/path/to/vagrantdir/.vagrant/machines/<name>/libvirt/private_key'` (of course if you're not using vagrant or you have another private key granted, then you should point at that instead)
```
Comment (jshimkus-rh) [inline Getting_Started.md]:
```suggestion
    NOTE: If using vagrant you can replace the use of `ansible_ssh_pass=vagrant` with `ansible_ssh_private_key_file='/path/to/vagrantdir/.vagrant/machines/<name>/libvirt/private_key'`.  If not using vagrant, or wanting to use a different private key, `ansible_ssh_private_key` should point to the appropriate private key.
```

File: Getting_Started.md (hunk 6 above)
Comment (jshimkus-rh) [inline Getting_Started.md]:
```suggestion
        - `--extra-vars="install_bunsen_support=true"` instructs the ansible playbook to install the necessary COPR repositories on the `controller` system and install the python dependencies (python3-resource-discovery, and python3-utility-mill).  It is required the first time you run the Bunsen playbook.  After that you may exclude using it unless you want to be certain to be using the latest version available; i.e., with it specified the most up-to-date version of the python dependencies will be installed.
```

File: Getting_Started.md (hunk 9 above)
Comment (rhawalsh) [reply Getting_Started.md:108]: Something else that does not exist yet.  That's what I was referring to in my comment above as hoping to write this week.

File: Getting_Started.md (hunk 1 above)
Comment (bjohnsto) [reply Getting_Started.md]: before moving on i think is what he wants there

Comment (bjohnsto) [reply Getting_Started.md]: not sure I like beginning. Beginning what?

File: Getting_Started.md (hunk 4 above)
Comment (bjohnsto) [reply Getting_Started.md]: I don't know about this one. Isn't is sort of important to have set up the "systems" beforehand with some knowledge of what they will be used for? I'm thinking of needing disk space for the farms. Seems like just saying 3 or 5 here doesn't really give them much to go on. Maybe we need to be more specific ?

File: Getting_Started.md (hunk 5 above)
Comment (bjohnsto) [reply Getting_Started.md]: This is the info I was talking about above. Maybe we don't need two places to discuss system count. Just here is enough?

**Review by bjohnsto — COMMENTED** [review]
I added some in line comments

File: Getting_Started.md (hunk 1 above)
Comment (rhawalsh) [reply Getting_Started.md]: The suggestion below from @jshimkus-rh is closer to what I had in my mind when writing this.

I will provide my response in both threads for consideration.

```suggestion
This project is an ansible playbook with specific requirements that need to be satisfied before it can be used to provision an environment.
```

Comment (rhawalsh) [reply Getting_Started.md]: I will provide my response in both threads for consideration.

```suggestion
This project is an ansible playbook with specific requirements that need to be satisfied before it can be used to provision an environment.
```

File: Getting_Started.md (hunk 2 above)
Comment (rhawalsh) [reply Getting_Started.md]: This will be in my next rev.

File: Getting_Started.md (hunk 3 above)
Comment (rhawalsh) [reply Getting_Started.md]: This will be in my next rev.

File: Getting_Started.md (hunk 4 above)
Comment (rhawalsh) [reply Getting_Started.md]: Thank you.  I was having trouble with this one.

This will be in my next rev.

File: Getting_Started.md (hunk 5 above)
Comment (rhawalsh) [reply Getting_Started.md]: This will be in my next rev.

File: Getting_Started.md (hunk 8 above)
Comment (rhawalsh) [reply Getting_Started.md]: This will be in my next rev.

File: Getting_Started.md (hunk 10 above)
Comment (rhawalsh) [reply Getting_Started.md]: This will be in my next rev.

File: Getting_Started.md (hunk 11 above)
Comment (rhawalsh) [reply Getting_Started.md]: This will be in my next rev.

File: Getting_Started.md (hunk 12 above)
Comment (rhawalsh) [reply Getting_Started.md]: This will be in my next rev.

File: Getting_Started.md (hunk 1 above)
Comment (jshimkus-rh) [reply Getting_Started.md]: Good by me.

File: Getting_Started.md (hunk 13 above)
Comment (rhawalsh) [reply Getting_Started.md]: This will be in my next rev.

File: Getting_Started.md (hunk 14 above)
Comment (rhawalsh) [reply Getting_Started.md]: This will be in my next rev.

File: Getting_Started.md (hunk 15 above)
Comment (rhawalsh) [reply Getting_Started.md]: This will be in my next rev.

File: Getting_Started.md (hunk 16 above)
Comment (rhawalsh) [reply Getting_Started.md]: A form of this will be in my next rev.  (probably without the trailing ...'s)

File: Getting_Started.md (hunk 17 above)
Comment (rhawalsh) [reply Getting_Started.md]: This will be in my next rev.

File: Getting_Started.md (hunk 18 above)
Comment (rhawalsh) [reply Getting_Started.md]: This will be in my next rev.

File: Getting_Started.md (hunk 19 above)
Comment (rhawalsh) [reply Getting_Started.md]: This will be in my next rev.

File: Getting_Started.md (hunk 6 above)
Comment (rhawalsh) [reply Getting_Started.md]: This will be in my next rev.

Comment (rhawalsh) [reply Getting_Started.md]: This will be in my next rev.

File: Getting_Started.md (hunk 7 above)
Comment (rhawalsh) [reply Getting_Started.md]: This will be in my next rev.

File: Getting_Started.md (hunk 1 above)
Comment (slegendr) [reply Getting_Started.md]: Good by me, as well.

**Review by slegendr — APPROVED** [review]
Approved for my part.

**Review by bjohnsto — APPROVED** [review]
Next rev sounds fine

File: Getting_Started.md [hunk 20]
```diff
@@ -1,47 +1,51 @@
...
 
-1. An environment must consist of at least 3 different systems (physical or virtual).  It is recommended that you have at least 5 machines, but minimally 3 machines can work.  More information on what the different systems will be doing will be provided later.
-    - Of those 3 machines, one must have at least 280GiB of additional storage in a separate block device (i.e. sdb)
-2. Each machine must be able to communicate with the others in the environment both by IP address and by name (i.e. ossbunsen-resource [10.2.0.5] needs to be able to ssh to ossbunsen-farm-f36-1 [10.2.0.6])
+1. At least 3 individual systems, physical and/or virtual.
+    - It is recommended to have at least 5 systems.
+    - The function of the separate systems are described below.
+    - At least one of the 3 systems must have a minimum of 280 GiB of additional storage in a separate block device.
+        - Generally this is /dev/sdb.
+        - If the system is virtualized using 'kvm', it can be /dev/vdb.
```
Comment (jshimkus-rh) [inline Getting_Started.md:18]: Bunsen file `provisioning/roles/farm/tasks/storage.yml` has the following:
```
      # Use the auxiliary disk, if it exists.
      - name: Setting scratch volume group and test disk names
        set_fact:
          scratch_vg: scratch
          test_disk: "/dev/{{aux_disk}}"

      vars:
        aux_disk: "{%- if ansible_virtualization_type == 'kvm' -%}
                     vdb
                   {%- else -%}
                     sdb
                   {%- endif -%}"
      when: aux_disk in ansible_devices.keys()
```

which is why my previous comment used "must" be /dev/vdb.  Based on the scenarios for setting `test_disk` (unless I missed something) it appears that if it's kvm and not vdb the provisioning will fail in
```
  - name: Verifying storage variables have been set
    assert:
      that:
        - scratch_vg is defined
        - test_disk is defined
```

Our general experience has been that the various virtualization technologies used automatically, in a two disk system, use either sdb or vdb.  Bunsen was written with that in mind.  Maybe it'd be better to make the description be that it must be a two drive system with the second disk being 280GiB, etc. rather than specifically mentioning device names.

File: Getting_Started.md [hunk 21]
```diff
@@ -1,47 +1,51 @@
...
 
 ## System Functions
 
-A provisioned Bunsen Environment consists of three different types of machine.
+A provisioned Bunsen Environment consists of three different classes of system.
 
 - Infrastructure
-    - This machine will provide necessary infrastructure services for the other systems to utilize.
+    - This system provides necessary services for the other systems to utilize.
         - Such as NFS, [RSVP](https://github.com/dm-vdo/permabit-rsvpd) (resource management and coordination), iSCSI backing storage (if applcable)
```
Comment (jshimkus-rh) [inline Getting_Started.md]:
```suggestion
        - Such as NFS, [RSVP](https://github.com/dm-vdo/permabit-rsvpd) (resource management and coordination), iSCSI backing storage (if applicable)
```

File: Getting_Started.md (hunk 20 above)
Comment (rhawalsh) [reply Getting_Started.md:18]: When I was calling out a particular device (sdb), it was really just to convey that it was another device, not the driver that was providing it.  Similarly, if you're on kvm, you can still use a different driver for the storage and end up with a `sdb` instead of `vdb`, but it's typical to get `vdb` if you're not messing with things.

I think I'd prefer the current spelling where we say it's generally going to be `/dev/sdb`, and if it's kvm that it can be `vdb`.  Perhaps I can rephrase it to mention using the `virtio driver` rather than kvm.

What do you think?

File: Getting_Started.md (hunk 21 above)
Comment (rhawalsh) [reply Getting_Started.md]: This is fixed.

**Comment by rhawalsh** [comment]
I am going to go through and resolve all of the outstanding conversations from people who have approved already to try and clean up the PR.  Please re-open any threads that you want to re-raise or think I made a mistake.

File: Getting_Started.md (hunk 6 above)
Comment (rhawalsh) [reply Getting_Started.md]: This is addressed in my recent pushes.

File: Getting_Started.md (hunk 7 above)
Comment (rhawalsh) [reply Getting_Started.md]: This is addressed in my recent pushes.

File: Getting_Started.md (hunk 9 above)
Comment (rhawalsh) [reply Getting_Started.md:108]: This will be addressed at a later time (maybe tonight?) :)

File: Getting_Started.md (hunk 1 above)
Comment (rhawalsh) [reply Getting_Started.md]: This is addressed in my recent pushes.

File: Getting_Started.md (hunk 3 above)
Comment (rhawalsh) [reply Getting_Started.md]: This is addressed in my recent pushes.

File: Getting_Started.md (hunk 4 above)
Comment (rhawalsh) [reply Getting_Started.md]: This is addressed in my recent pushes.

File: Getting_Started.md (hunk 5 above)
Comment (rhawalsh) [reply Getting_Started.md]: This is addressed in my recent pushes.

File: Getting_Started.md (hunk 8 above)
Comment (rhawalsh) [reply Getting_Started.md]: This is addressed in my recent pushes.

File: Getting_Started.md (hunk 10 above)
Comment (rhawalsh) [reply Getting_Started.md]: This is addressed in my recent pushes.

File: Getting_Started.md (hunk 11 above)
Comment (rhawalsh) [reply Getting_Started.md]: This is addressed in my recent pushes.

File: Getting_Started.md (hunk 13 above)
Comment (rhawalsh) [reply Getting_Started.md]: This is addressed in my recent pushes.

File: Getting_Started.md (hunk 14 above)
Comment (rhawalsh) [reply Getting_Started.md]: This is addressed in my recent pushes.

File: Getting_Started.md (hunk 15 above)
Comment (rhawalsh) [reply Getting_Started.md]: This is addressed in my recent pushes.

File: Getting_Started.md (hunk 16 above)
Comment (rhawalsh) [reply Getting_Started.md]: This is addressed in my recent pushes.

File: Getting_Started.md (hunk 18 above)
Comment (rhawalsh) [reply Getting_Started.md]: This is addressed in my recent pushes.

File: Getting_Started.md (hunk 19 above)
Comment (rhawalsh) [reply Getting_Started.md]: This is addressed in my recent pushes.

File: Getting_Started.md (hunk 6 above)
Comment (rhawalsh) [reply Getting_Started.md]: This is addressed in my recent pushes.

File: Getting_Started.md (hunk 2 above)
Comment (rhawalsh) [reply Getting_Started.md]: This is addressed in my recent pushes.

File: Getting_Started.md (hunk 12 above)
Comment (rhawalsh) [reply Getting_Started.md]: This is addressed in my recent pushes.

File: Getting_Started.md (hunk 20 above)
Comment (jshimkus-rh) [reply Getting_Started.md:18]: Simply having a second storage device is not sufficient for it to be used; for example, I can create a second storage device as NVMe and it's not touched.  The current implementation of bunsen has a number of fallbacks but they're generally predicated on there being enough free space on the root disk; if the root disk doesn't have enough free space provisioning will fail.

There are, as I see it, three choices:
1. extend bunsen to handle more scenarios w/o requiring specific device names
2. proscribe that in environment X the second device must be named Y
3. rely on our experiences that "the right thing" happens and do not mention any device name

The robustness of option 1 will probably always have some holes.
Option 2 is akin to any other environmental qualification; it might not meet some folks desires but they could then extend bunsen.
Option 3, if someone encountered a problem, could be explained away as we've been using default behavior of various virtualization tech; if they are specifying a non-default behavior that's their issue.  Again, they could extend bunsen.

The engineer in me would like to do option 1, but pragmatically I think either option 2 or 3 is the way to go at present.

Comment (rhawalsh) [reply Getting_Started.md:18]: I think we need to hit the pause button on this discussion.

The primary purpose of this PR is to establish a set of instructions to  use; something we don't have posted yet.

So with all that being said, I think we should go with the easiest option and not be too strict about our spelling.  We can always adjust it after the fact.  If we are going to decide only through the three options you provided, then given these constraints, I would opt for option 3 and just omit the names and let people figure it out the hard way until someone with more time can address it more completely.

**Review by jshimkus-rh — APPROVED** [review]
